File: goal/src/main/java/dev/exit/asset/goal/ExitGoal.java
package dev.exit.asset.goal;

import java.math.BigDecimal;
import java.time.LocalDate;

import dev.exit.asset.common.OwnedEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;

@Entity
@Table(name = "exit_goals")
public class ExitGoal extends OwnedEntity {

	@Column(nullable = false, length = 120)
	private String name;

	@Column(nullable = false, precision = 19, scale = 2)
	private BigDecimal targetNetWorth;

	@Column(nullable = false)
	private LocalDate targetDate;

	@Column(nullable = false, precision = 5, scale = 2)
	private BigDecimal targetSavingsRate;

	@Column(nullable = false)
	private boolean active;

	protected ExitGoal() {
	}

	public ExitGoal(String ownerId, String name, BigDecimal targetNetWorth, LocalDate targetDate,
			BigDecimal targetSavingsRate, boolean active) {
		super(ownerId);
		this.name = name;
		this.targetNetWorth = targetNetWorth;
		this.targetDate = targetDate;
		this.targetSavingsRate = targetSavingsRate;
		this.active = active;
	}

	public String getName() {
		return name;
	}

	public BigDecimal getTargetNetWorth() {
		return targetNetWorth;
	}

	public LocalDate getTargetDate() {
		return targetDate;
	}

	public BigDecimal getTargetSavingsRate() {
		return targetSavingsRate;
	}

	public boolean isActive() {
		return active;
	}
}

File: goal/src/main/kotlin/dev/exit/asset/goal/ExitGoal.kt
package dev.exit.asset.goal

import dev.exit.asset.common.OwnedEntity
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Table
import java.math.BigDecimal
import java.time.LocalDate

@Entity
@Table(name = "exit_goals")
open class ExitGoal protected constructor() : OwnedEntity() {

    @Column(nullable = false, length = 120)
    lateinit var name: String
        protected set

    @Column(nullable = false, precision = 19, scale = 2)
    lateinit var targetNetWorth: BigDecimal
        protected set

    @Column(nullable = false)
    lateinit var targetDate: LocalDate
        protected set

    @Column(nullable = false, precision = 5, scale = 2)
    lateinit var targetSavingsRate: BigDecimal
        protected set

    @Column(nullable = false)
    var active: Boolean = false
        protected set

    constructor(
        ownerId: String,
        name: String,
        targetNetWorth: BigDecimal,
        targetDate: LocalDate,
        targetSavingsRate: BigDecimal,
        active: Boolean,
    ) : this() {
        this.ownerId = ownerId
        this.name = name
        this.targetNetWorth = targetNetWorth
        this.targetDate = targetDate
        this.targetSavingsRate = targetSavingsRate
        this.active = active
    }
}
